arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -0,0 +1,1563 @@
+
+          // if the server is still in the list, it was neither in plan nor in current
+          if (isResponsible) {
+
+            auto rootPath = arangodb::cluster::paths::root()->arango();
+            auto planVersionPath = rootPath->plan()->version();
+            // do a write transaction if server is no longer used
+            VPackBuffer<uint8_t> trx;
+            {
+              VPackBuilder builder(trx);
+              arangodb::agency::envelope<VPackBuilder>::create(builder)
+                .write()
+                  .remove(rootPath->plan()->coordinators()->server(ctx->server)->str())
+                  .remove(rootPath->plan()->dBServers()->server(ctx->server)->str())
+                  .remove(rootPath->current()->serversRegistered()->server(ctx->server)->str())
+                  .remove(rootPath->current()->dBServers()->server(ctx->server)->str())
+                  .remove(rootPath->supervision()->health()->server(ctx->server)->str())
+                  .remove(rootPath->target()->mapUniqueToShortID()->server(ctx->server)->str())
+                  .remove(rootPath->current()->serversKnown()->server(ctx->server)->str())
+                  .set(rootPath->target()->removedServers()->server(ctx->server)->str(),
+                    timepointToString(std::chrono::system_clock::now()))
+                .precs()
+                  .isEqual(rootPath->supervision()->health()->server(ctx->server)->status()->str(), "FAILED")
+                  .isEmpty(rootPath->supervision()->dbServers()->server(ctx->server)->str())
+                  .isEqual(planVersionPath->str(), agency.get(planVersionPath->vec()))
+                .done()
+              .done();
+            }
+
+            return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+              [this, ctx = std::move(ctx)](AsyncAgencyCommResult &&result) mutable {
+                if (result.ok()) {
+                  if (result.statusCode() == fuerte::StatusOK) {
+                    resetResponse(rest::ResponseCode::OK);
+                    return futures::makeFuture();
+                  } else if (result.statusCode() == fuerte::StatusPreconditionFailed) {
+                    return retryTryDeleteServer(std::move(ctx));
+                  }
+                }
+                generateError(result.asResult());
+                return futures::makeFuture();
+            });
+          }
+
+          return retryTryDeleteServer(std::move(ctx));
+        } else {
+          generateError(rest::ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND);
+        }
+      } else {
+        generateError(result.asResult());
+      }
+
+      return futures::makeFuture();
+    });
+}
+
+RestStatus RestAdminClusterHandler::handlePostRemoveServer(std::string const& server) {
+  auto ctx = std::make_unique<RemoveServerContext>(server);
+
+  return waitForFuture(tryDeleteServer(std::move(ctx))
+    .thenError<VPackException>([this](VPackException const& e) {
```

Comment:
does this need to capture `self`?

Reply:
This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.